rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+`version-directive` resource.
+
+- *version reconciliation loop*: A control loop that runs in the auth server which
+compares the desired state as specified by the `version-directive` with the current
+state of teleport installations across the cluster. When mismatches are discovered,
+the appropriate `installer`s are run.
+
+- `installer`: A component capable of attempting to affect installation
+of a specific target on one or more teleport hosts. The auth server needs to know
+enough to at least start the process, but the core logic of a given installer
+may be an external command (e.g. the proposed `local-script` installer would cause
+each teleport instance in need of upgrade to run a user-supplied script locally and
+then restart). From the perspective of a user, an installer is a teleport
+configuration object (though that configuration object may only be a thin hook into
+the "real" `installer`). Whether or not the teleport instance being upgraded
+needs to understand the installer will vary depending on type.
+
+There is room in the above model for additional granularity (e.g. schedulers), but
+the listed components are a decent starting point.
+
+
+### Implementation Phases
+
+Implementation will be divided into a series of phases consisting of 1 or more
+separate PRs/releases each:
+
+- Setup: Changes required for inventory status/control model, but not necessarily
+specific to the upgrade system.
+
+- Notification-Only System: Optional phase intended to deliver value sooner
+at the cost of overall feature development time.
+
+- Script-Based Installs MVP: Early MVP supporting only manual control and
+simple script-based installers for non-auth instances.
+
+- TUF-Based System MVP: Fully functional but minimalistic upgrade system
+based on TUF (still excludes auth instances).
+
+- Stability & Polish: Additional necessary features (including auth installs
+and local rollbacks). Represents the point at which the core upgrade system
+can be considered "complete".
+
+- Extended Feature Set: A collection of nice-to-haves that we're pretty sure
+folks are going to want.
+
+See the [implementation plan](#implementation-plan) section for detailed breakdown
+of the required elements for each phase.
+
+
+## Details
+
+### Usage Scenarios
+
+
+#### Notification-Only Usecase
+
+Cluster administrators may or may not be using teleport's own install mechanisms, but they
+want teleport to be able to inform them when instances are outdated, and possibly generate
+alerts if teleport is running a deprecated version and/or there is a newer security patch
+available.
+
+In this case we want options for both displaying suggested version (or just a "needs upgrade"
+badge) on inventory views, and also probably some kind of cluster-wide alert that can be
+made difficult to ignore (e.g. a message on login or a banner in the web UI). We also
+probably want an API that will support plugins that can emit alerts to external locations
```

Comment:
@fspmarshall I think adding a state to our prometheus endpoint and possibly `/secure` endpoint with `200` and non 200 endpoint for scripting will be super helfpul